stablehlo/reference/Element.cpp

```diff
@@ -607,6 +601,203 @@ Element ceil(const Element &el) {
+                    /*isSigned=*/isSupportedSignedIntegerType(type)));
+  if (isSupportedFloatType(type)) {
+    auto floatVal = APFloat(static_cast<double>(value));
+    bool roundingErr;
+    floatVal.convert(type.cast<FloatType>().getFloatSemantics(),
+                     APFloat::rmNearestTiesToEven, &roundingErr);
+    return Element(type, floatVal);
+  }
+  if (isSupportedComplexType(type)) {
+    const llvm::fltSemantics &fltSemantics = type.cast<ComplexType>()
+                                                 .getElementType()
+                                                 .cast<FloatType>()
+                                                 .getFloatSemantics();
+    bool roundingErr;
+    auto realVal = APFloat(static_cast<double>(value));
+    realVal.convert(fltSemantics, APFloat::rmNearestTiesToEven, &roundingErr);
+    auto imagVal = APFloat(0.0);
+    imagVal.convert(fltSemantics, APFloat::rmNearestTiesToEven, &roundingErr);
+    return Element(type, std::complex<APFloat>(realVal, imagVal));
+  }
+  report_fatal_error(invalidArgument("Unsupported element type: %s",
+                                     debugString(type).c_str()));
+}
+
+Element convert(Type type, uint64_t value) {
+  if (isSupportedBooleanType(type)) return Element(type, value != 0);
+  if (isSupportedIntegerType(type))
+    return Element(type,
+                   APInt(type.getIntOrFloatBitWidth(), value,
+                         /*isSigned=*/isSupportedSignedIntegerType(type)));
+  if (isSupportedFloatType(type)) {
+    bool roundingErr;
+    auto floatVal = APFloat(static_cast<double>(value));
+    floatVal.convert(type.cast<FloatType>().getFloatSemantics(),
+                     APFloat::rmNearestTiesToEven, &roundingErr);
+    return Element(type, floatVal);
+  }
+  if (isSupportedComplexType(type)) {
+    const llvm::fltSemantics &fltSemantics = type.cast<ComplexType>()
+                                                 .getElementType()
+                                                 .cast<FloatType>()
+                                                 .getFloatSemantics();
+    bool roundingErr;
+    auto realVal = APFloat(static_cast<double>(value));
+    realVal.convert(fltSemantics, APFloat::rmNearestTiesToEven, &roundingErr);
+    auto imagVal = APFloat(0.0);
+    imagVal.convert(fltSemantics, APFloat::rmNearestTiesToEven, &roundingErr);
+    return Element(type, std::complex<APFloat>(realVal, imagVal));
+  }
+  report_fatal_error(invalidArgument("Unsupported element type: %s",
+                                     debugString(type).c_str()));
+}
+
+Element convert(Type type, APFloat value) {
+  if (isSupportedBooleanType(type)) return Element(type, value.isNonZero());
+  if (isSupportedIntegerType(type))
+    return Element(
+        type, llvm::APIntOps::RoundDoubleToAPInt(value.convertToDouble(),
+                                                 type.getIntOrFloatBitWidth()));
```

Comment:
Similarly to the comment on the int64_t overload, can we use `APFloat::convertToInteger` instead to keep all the conversions in the APInt/APFloat domain?

Reply:
Not sure if you were aware of this, but one important thing I noticed is that by using `APFloat::convertToInteger`, we now explicitly write the rounding mode from float->int (i.e. `rmTowardZero`, which is different from the general RM of the codebase `rmNearestTiesToEven`). This brings the implementation more aligned with the spec :)